cpp/include/cudf/detail/copy_range.cuh

```diff
@@ -0,0 +1,308 @@
+        atomicAdd(null_count, block_null_change);
+      }
+    }
+  }
+}
+
+}  // namespace anonymous
+
+namespace cudf {
+namespace experimental {
+namespace detail {
+
+/**
+ * @brief Internal API to copy a range of values from a functor to a column.
+ *
+ * Copies N values from @p source to the range [@p begin, @p end)
+ * of @p target. @p target is modified in place.
+ *
+ * SourceFunctor must have these accessors:
+ * __device__ T data(cudf::size_type index);
+ * __device__ bool valid(cudf::size_type index);
+ *
+ * @tparam SourceFunctor the type of the source function object
+ * @param source An instance of SourceFunctor that provides data and valid mask
+ * @param target the column to copy into
+ * @param begin The beginning of the target range to write to
+ * @param end The index after the last element of the target range to write to
+ * @param stream CUDA stream to run this function
+ */
+template <typename SourceValueIterator, typename SourceValidityIterator>
+void copy_range(SourceValueIterator source_value_begin,
+                SourceValidityIterator source_validity_begin,
+                mutable_column_view& target,
+                size_type target_begin, size_type target_end,
+                cudaStream_t stream = 0) {
+  CUDF_EXPECTS((target_begin <= target_end) &&
+                 (target_begin >= 0) &&
+                 (target_begin < target.size()) &&
+                 (target_end <= target.size()),
+               "Range is out of bounds.");
+  using T = typename std::iterator_traits<SourceValueIterator>::value_type;
+
+  // this code assumes that source and target have the same type.
+  CUDF_EXPECTS(type_to_id<T>() == target.type().id(), "the data type mismatch");
+
+  static_assert(warp_size == cudf::util::size_in_bits<bitmask_type>(),
+    "copy_range_kernel assumes bitmask element size in bits == warp size");
+
+  T * __restrict__ data = target.head<T>();
+  bitmask_type * __restrict__ bitmask = target.null_mask();
+  auto offset = target.offset();
+
+#if 1
+  auto warp_aligned_begin_lower_bound =
+    size_type{target_begin - (target_begin % warp_size)};
+  auto warp_aligned_end_upper_bound =
+    size_type{
+      (target_end % warp_size) == 0 ? \
+        target_end : target_end + (warp_size - (target_end % warp_size))
+    };
+  auto num_items =
+    warp_aligned_end_upper_bound - warp_aligned_begin_lower_bound;
+#else
+  // This one results in a compiler internal error! TODO: file NVIDIA bug
+  // size_type num_items =
+  //   cudf::util::round_up_safe(target_end - target_begin, warp_size);
+  // number threads to cover range, rounded to nearest warp
+  // this code runs for one additional round if target_begin is not warp
+  // aligned, and target_end is block_size + 1
+  auto num_items =
+    size_type{
+      warp_size *
+        cudf::util::div_rounding_up_safe(target_end - target_begin, warp_size)
+    };
+#endif
+
+  constexpr auto block_size = int{256};
+  static_assert(block_size <= 1024,
```

Comment:
Inside the kernel, blockDim.x cannot be evaluated in compile time, so I need to do run-time check instead of compile-time static_assert, do you still want me to move this to the kernel?

Reply:
I thought the `block_size` was a template param of the kernel.

Reply:
With your suggestion to use _single_lane_block_sum_reduce_, this indeed becomes a template parameter, and this is checked inside the single_lane_block_sum_reduce, so this discussion may become irrelevant; actually _single_lane_block_sum_reduce_ does what this code is doing, so, I am updating this code to use _single_lane_block_sum_reduce_ to avoid reimplementing the same logic.